kube-proxy e2e  tests

ArtfulCoder opened the pull request.

ArtfulCoder committed:
kube-proxy test

thockin commented on test/images/netproxy/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+FROM debian:wheezy
```
why wheezy rather than busybox?

ArtfulCoder replied:
no real reason. changed.

thockin commented on test/images/netproxy/netproxy.go:
```diff
@@ -0,0 +1,228 @@
+	host := values.Query().Get("host")
+	port := values.Query().Get("port")
+	request := values.Query().Get("request") // hostName
+	protocol := values.Query().Get("protocol")
```
is this "tcp/udp" or "http/https"?  If the latter, maybe "scheme" is a better name

thockin replied:
I see below, nevermind

thockin commented on test/images/netproxy/netproxy.go:
```diff
@@ -0,0 +1,228 @@
+		return
+	}
+
+	hostPort := fmt.Sprintf("%s:%s", host, port)
```
net.JoinHostPort ?

what if port is ""?

ArtfulCoder replied:
added ResolveTCPAddr check
using JoinHostPort

thockin commented on test/images/netproxy/netproxy.go:
```diff
@@ -0,0 +1,228 @@
+		n, clientAddress, err := serverConn.ReadFromUDP(buf)
+		assertNoError(err)
+		receivedText := strings.TrimSpace(string(buf[0:n]))
+		if receivedText == "hostName" {
```
so UDP only supports hostname but HTTP also supports "shell" and "dial"?  Just being clear - maybe a comment?

ArtfulCoder replied:
user gets a response saying unknown udp command.
I can add a comment too

thockin commented on test/e2e/kubeproxy.go:
```diff
@@ -0,0 +1,447 @@
+	nodeHttpPort            = 32080
+	nodeUdpPort             = 32081
+	loadBalancerHttpPort    = 100
+	imageName               = "artfulcoder/netproxy:1.0"
```
BEFORE MERGE: we'll need to push this to gcr.io/google_containers

thockin commented on test/e2e/kubeproxy.go:
```diff
@@ -0,0 +1,447 @@
+		By("cleaning up any pre-existing namespaces used by this test")
+		f.cleanup()
+
+		By("creating a selector")
```
@ixdy  Is this appropriate use of By() ?  I thought By() was test cases, not just progress marks...

quinton-hoole replied:
No, By() is just progress marks. http://onsi.github.io/ginkgo/#documenting-complex-its-by

thockin commented on test/e2e/kubeproxy.go:
```diff
@@ -0,0 +1,447 @@
+	//testContext.CloudConfig.NumNodes
+	pods := make([]*api.Pod, 0)
+
+	for i := 0; i < nodeCount; i++ {
```
Why not a replication controller?  If there's a good reason it should be commented here

thockin replied:
Should we be assigning one backend pod per node?

thockin commented on test/e2e/kubeproxy.go:
```diff
@@ -0,0 +1,447 @@
+	By("dialing(udp) node1:nodeUdpPort from container")
+	config.dialFromContainer("udp", node1_IP, nodeUdpPort, tries, epCount)
+	By("dialing(http) node1:nodeHttpPort from container")
+	// TODO:this test seems to expose an issue. unbalanced
```
Is this TODO still open?  What's the issue?

thockin commented:
This is a really nice test.  Mostly nits.


Can you add a TODO to test access to externalIPs on the service port?

We should also add a test that kills (but doesn't restart) the proxy and verifies connections then restarts the proxy and verifies connections.

We should also add a test that deletes a service and verifies that connections fail

I will accept TODOs for all of these to get this in so we can do more manual tests

ArtfulCoder commented:
added delete-a-service test
added hairpin test

TODO: externalIPs test (will do it in a seperate PR)
We do have a seperate test for externalIPs that I had written and is checked in, but not part of this particular test

Please review

ArtfulCoder commented:
The e2e failure is unrelated.
The KubeProxy test was successful.

thockin commented on test/images/netproxy/Makefile:
```diff
@@ -0,0 +1,15 @@
+	CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -ldflags '-w' ./netproxy.go
+
+image: netproxy
+	docker build -t gcr.io/google_containers/netproxy:$(TAG) .
```
make a variable, usually called PREFIX for "gcr.io/google_containers"

thockin commented on test/images/netproxy/netproxy.go:
```diff
@@ -0,0 +1,235 @@
+limitations under the License.
+*/
+
+package main
```
this whole dir should go in contrib/for_tests, and probably needs a better name.  netexec?

thockin replied:
We could maybe get rid of serve_hostname in our codebase in favor of this?

ArtfulCoder replied:
I called it netproxy, because we can use it to exec and also dial other endpoints through it.
So it acts as a proxy to shell or other endpoints..

are you concerned that is sounds similar to kubeproxy ?

thockin replied:
yeah "netproxy" is just very generic.

On Mon, Sep 14, 2015 at 10:20 PM, Abhi Shah <notifications@github.com>
wrote:

> In test/images/netproxy/netproxy.go
> <https://github.com/kubernetes/kubernetes/pull/13697#discussion_r39475374>
> :
>
> > +Copyright 2014 The Kubernetes Authors All rights reserved.
> > +
> > +Licensed under the Apache License, Version 2.0 (the "License");
> > +you may not use this file except in compliance with the License.
> > +You may obtain a copy of the License at
> > +
> > +    http://www.apache.org/licenses/LICENSE-2.0
> > +
> > +Unless required by applicable law or agreed to in writing, software
> > +distributed under the License is distributed on an "AS IS" BASIS,
> > +WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > +See the License for the specific language governing permissions and
> > +limitations under the License.
> > +*/
> > +
> > +package main
>
> I called it netproxy, because we can use it to exec and also dial other
> endpoints through it.
> So it acts as a proxy to shell or other endpoints..
>
> are you concerned that is sounds similar to kubeproxy ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13697/files#r39475374>.
>

thockin commented on test/e2e/kubeproxy.go:
```diff
@@ -0,0 +1,487 @@
+	})
+})
+
+func (config *KubeProxyTestConfig) hit() {
```
rename to testAl() ?

ArtfulCoder replied:
maybe call it hitAll() to match other method names like hitEndpoints() ?

thockin commented:
This is a really nice test.

LGTM after minor points.  I think you'll have to break the tester to a new PR, commit that, push the image, then finish this.

ArtfulCoder commented:
I am using netexec.
I renamed hit to hitall.
If this e2e run is green for the added test, we should be good to go.

alex-mohr commented:
Test failures for jenkins are "Kubernetes e2e suite.KubeProxy should test kube-proxy" so seems likely this PR needs updating?

ArtfulCoder commented:
yes, I will investigate this by running it manually.

ArtfulCoder commented:
test was green. the failing test is not the one introduced in this PR.

alex-mohr commented:
@k8s-bot test this please (previous fail #14076)

ixdy commented:
Not sure how this passed SQ, because it's breaking things now:
```
15:39:37 # k8s.io/kubernetes/test/e2e
15:39:37 _output/local/go/src/k8s.io/kubernetes/test/e2e/kubeproxy.go:275: undefined: "k8s.io/kubernetes/pkg/api".TypeMeta
15:39:37 _output/local/go/src/k8s.io/kubernetes/test/e2e/kubeproxy.go:318: undefined: "k8s.io/kubernetes/pkg/api".TypeMeta
```

ixdy commented:
ah, it was a race with #14137 which did not go through the SQ. @alex-mohr @caesarxuchao

alex-mohr commented:
Sorry -- was trying to get #14137 in before it had to be rebased a 10th time.  Trying to get a fix pr sent..

mvdan commented:
Ah! This just bit me too.
